### DISCUSSION TITLE
Message “package … does not exist” despite of a corresponding Java source file

### DISCUSSION THREAD

**Discussion by Markus-Elfring** [apache/netbeans#4566] [Q&A]
### Apache NetBeans version

Apache NetBeans 14

### What happened

[The script “Apache-NetBeans-14-bin-linux-x64.sh”](https://netbeans.apache.org/download/nb14/nb14.html "Downloading Apache NetBeans 14") was successfully executed also on my development system.
I opened a test project a moment later.

I wondered then about the error message “package … does not exist”.

![questionable_message-NetBeans-excerpt-20220829](https://user-images.githubusercontent.com/102525885/187194233-cda867e4-da08-4ce0-bb39-cdf1826bb078.png)

### How to reproduce

Please take another look at import specifications (also in known Java example source files).

### Did this work correctly in an earlier version?

No / Don't know

### Operating System

Linux 5.19.2-1.2 x86_64 - openSUSE Tumbleweed

### JDK

OpenJDK 18.0.2.0 x86_64 - openSUSE Tumbleweed

### Apache NetBeans packaging

Apache NetBeans provided installer

### Anything else

The software “Eclipse IDE for Java Developers 2022-06 (4.24.0)” found the mentioned import statement usable.

-----

Another example for a previous clarification attempt:
2016-08-11: ["Package does not exist" but it does](https://bz.apache.org/netbeans/show_bug.cgi?id=267538 "Clarification approach")

### Are you willing to submit a pull request?

No

### Code of Conduct

Yes

**Comment by neilcsmith-net** [comment]
That's not a valid project. The packages should be under `src/main/java`.

**Comment by Chris2011** [comment]
> That's not a valid project. The packages should be under `src/main/java`.

Does this really matters and does this really make a difference?

**Comment by Markus-Elfring** [comment]
>The packages should be under `src/main/java`.

:thought_balloon: I suggest to reconsider such information a bit more also according to a required statement like “`package picocli;`” (for my test example).
Do development environments present differences in the desired storage locations?

**Comment by neilcsmith-net** [comment]
> Does this really matters and does this really make a difference?

Yes. @Chris2011 if it's not a valid project how do you expect NetBeans to work out where the root of Java packages is?

> Do development environments present differences in the desired storage locations?

Well, build systems do, and unless the build system (eg. Maven) is configured correctly, then it's neither going to build or allow NetBeans to find the required information.  I hadn't noticed initially, but it looks like the `pom.xml` file is incorrectly under `src` too.  This might accidentally work in Eclipse because of the `.classpath` file, but it's not going to work elsewhere.

For help with Maven inside (or outside) NetBeans, I suggest you ask on the users@ mailing list or in the GitHub discussions.

**Comment by Markus-Elfring** [comment]
>I hadn't noticed initially, but it looks like the `pom.xml` file is incorrectly under `src` too.

The file “pom.xml” is stored in the root directory for my test project example.
:thought_balloon: Do you expect a clearer graphical representation by the NetBeans IDE accordingly?

**Comment by Markus-Elfring** [comment]
>… where the root of Java packages is?

Does the shown screenshot excerpt indicate relevant information? :thinking:

**Reply by Markus-Elfring** [reply]
:thought_balloon: Would you like to reconsider software development consequences also according to information like the following?

The Java® Language Specification
Java SE 18 Edition
Section “[Fully Qualified Names and Canonical Names](https://docs.oracle.com/javase/specs/jls/se18/html/jls-6.html#jls-6.7 "Reference information")”:
“…
The fully qualified name of a named package that is not a subpackage of a named package is its simple name.
…
The fully qualified name of a top level class or top level interface that is declared in a named package consists of the fully qualified name of the package, followed by ".", followed by the simple name of the class or interface.
…”


Is it supported then (by available development tools) to store a package with a “simple name” in the directory “<project_root>/src” directly besides the known subdirectory “main”? :thinking:

**Reply by sdedic** [reply]
> > … where the root of Java packages is?
> 
> Does the shown screenshot excerpt indicate relevant information? thinking

No it does not. And if you worked with the "Project" tab, you would clearly see where the package root is, and that your `CommandLine.java` file is outside of that root - and therefore won't be considered by the compiler or parser for resolution.

What's interesting is the very existence of  `picocli/CommandLine.java` in your project's source tree: `picocli.CommandLine` is a class from picocli library, and as such should be included in compiled form in a library listed in the project's POM.  Are you sure your project is set up correctly ?

**Reply by Markus-Elfring** [reply]
>What's interesting is the very existence of `picocli/CommandLine.java` in your project's source tree: `picocli.CommandLine` is a class from picocli library,

I copied this source file from the published archive “picocli-4.6.3.tar.gz”.


>and as such should be included in compiled form in a library listed in the project's POM.

I came along other software build expectations for this component.
:thought_balloon: I would eventually like to bundle two source files for my test project example because of special circumstances.


>Are you sure your project is set up correctly ?

I got the impression that the corresponding answer depends on the selected development tool versions. :thinking:

**Comment by neilcsmith-net** [comment]
Moving this into discussions as suggested.
